src/prefect/futures.py

```diff
@@ -233,6 +235,60 @@ def __eq__(self, other):
         return self.task_run_id == other.task_run_id
 
 
+class PrefectFutureList(list, Iterator, Generic[F]):
+    """
+    A list of Prefect futures.
+
+    This class provides methods to wait for all futures
+    in the list to complete and to retrieve the results of all task runs.
+    """
+
+    def wait_for_all(self, timeout: Optional[float] = None) -> None:
+        """
+        Wait for all futures in the list to complete.
+
+        Args:
+            timeout: The maximum number of seconds to wait for all futures to
+                complete. This method will not raise if the timeout is reached.
+        """
+        try:
+            with timeout_context(timeout):
+                for future in self:
+                    future.wait()
+        except TimeoutError:
+            logger.debug("Timed out waiting for all futures to complete.")
+            return
+
+    def results(
+        self,
+        timeout: Optional[float] = None,
+        raise_on_failure: bool = True,
+    ) -> List:
+        """
+        Get the results of all task runs associated with the futures in the list.
+
+        Args:
+            timeout: The maximum number of seconds to wait for all futures to
+                complete.
+            raise_on_failure: If `True`, an exception will be raised if any task run fails.
+
+        Returns:
+            A list of results of the task runs.
+
+        Raises:
+            TimeoutError: If the timeout is reached before all futures complete.
+        """
+        try:
+            with timeout_context(timeout):
+                return [
+                    future.result(raise_on_failure=raise_on_failure) for future in self
+                ]
+        except TimeoutError as exc:
+            raise TimeoutError(
+                f"Timed out waiting for all futures to complete within {timeout} seconds"
+            ) from exc
```

Comment:
Ooo, this one is tricky with `raise_on_failure`...what if it was one of the _tasks_ that raised a `TimeoutError`?  We won't be able to distinguish between the two, will we?  

(definitely not something to block the PR over)

Reply:
I guess since you're raising the new exception `from` the inner one, you might be okay since the inner exception is still available

Reply:
Fixed in https://github.com/PrefectHQ/prefect/pull/14234/commits/e50312dfa171f1cf0e739a051b4eaaca3657c78f